subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java

```diff
@@ -57,6 +57,19 @@
      */
     <T> Provider<T> provider(Callable<? extends T> value);
 
+    /**
+     * Creates a caching {@link Provider} whose value is calculated using the given {@link Callable}.
+     *
+     * <p>The provider will call the {@link Callable} on the first query and will cache its value for subsequent queries. The {@link Callable} may return {@code null}, in which case the provider is considered to have no value.
+     *
+     * @param value The {@code java.util.concurrent.Callable} use to calculate the value.
+     * @return The provider. Never returns null.
+     *
+     * @since 7.5
+     */
+    @Incubating
+    <T> Provider<T> cachingProvider(Callable<? extends T> value);
```

Comment:
Yes, I agree with @melix that such a primitive should take a provider.

The implementation could be based on a `finalizeValueOnRead() + disallowChanges()` property instead of introducing a new provider type.

Reply:
I tried to use the existing property implementation as suggested and ran into a few issues:
1) Can I inject the `ObjectFactory` into the `DefaultProviderFactory` to get an instance of a property?
   - Or maybe it's safe to create it using the `DefaultProperty` constructor, i.e. `new DefaultProperty<>(PropertyHost.NO_OP, ...)`?
2) How do I create this property if the class of the object is not available at runtime?
   - I tried using `ProviderInternal#getType`, but that returns null for some providers (e.g. `DefaultProvider`).
   - Just using `Object.class` and doing an unchecked cast seems to work, but I'm not sure if that has any other side effects.